include/alpaka/atomic/AtomicUniformCudaHipBuiltIn.hpp

```diff
@@ -37,998 +30,301 @@ namespace alpaka
 #        if defined(ALPAKA_ACC_GPU_HIP_ENABLED) && !BOOST_LANG_HIP
 #            error If ALPAKA_ACC_GPU_HIP_ENABLED is set, the compiler has to support HIP!
 #        endif
-
+namespace alpaka
+{
     namespace trait
     {
-        //! The specializations to execute the requested atomic ops of the CUDA/HIP accelerator.
-        // See: http://docs.nvidia.com/cuda/cuda-c-programming-guide/#atomic-functions how to implement everything with
-        // CAS
-
-        // Add.
-
-        //! The GPU CUDA/HIP accelerator atomic operation.
-        template<typename THierarchy>
-        struct AtomicOp<AtomicAdd, AtomicUniformCudaHipBuiltIn, int, THierarchy>
+        namespace detail
         {
-            __device__ static auto atomicOp(AtomicUniformCudaHipBuiltIn const&, int* const addr, int const& value)
-                -> int
+            struct EmulationBase
             {
-                return ::atomicAdd(addr, value);
-            }
-        };
-        //! The GPU CUDA/HIP accelerator atomic operation.
-        template<typename THierarchy>
-        struct AtomicOp<AtomicAdd, AtomicUniformCudaHipBuiltIn, unsigned int, THierarchy>
-        {
-            __device__ static auto atomicOp(
-                AtomicUniformCudaHipBuiltIn const&,
-                unsigned int* const addr,
-                unsigned int const& value) -> unsigned int
-            {
-                return ::atomicAdd(addr, value);
-            }
-        };
-        //! The GPU CUDA/HIP accelerator atomic operation.
-        template<typename THierarchy>
-        struct AtomicOp<AtomicAdd, AtomicUniformCudaHipBuiltIn, unsigned long int, THierarchy>
-        {
-            //
-            __device__ static auto atomicOp(
-                AtomicUniformCudaHipBuiltIn const&,
-                unsigned long int* const addr,
-                unsigned long int const& value) -> unsigned long int
-            {
-                // LLP64
-                if constexpr(std::numeric_limits<unsigned int>::max() == std::numeric_limits<unsigned long>::max())
-                    return ::atomicAdd(reinterpret_cast<unsigned int*>(addr), static_cast<unsigned int>(value));
-                else // LP64
+                //! reinterprets an adress to an 32bit value for atomicCas emulation usage
+                template<typename TAddressType>
+                __device__ static auto reinterpretAddress(TAddressType* address)
+                    -> std::enable_if_t<sizeof(TAddressType) == 4u, unsigned int*>
```

Comment:
Should a fixed-width type be used instead of `unsigned int`? Then the 32-bit requirement is also more obvios. However I am not sure if it bodes well with CUDA and HIP.

Reply:
No, we need to use a type where we know that the signature fits to CUDA/HIP. If we use `std::uint32_t` is can link to `unsigned long int` or `unsigned int` depending on the platform.

Reply:
Okay, but then we should also check that `sizeof(unsigned int) == 4` ?

Reply:
Discussed offline: maybe it's not worth as nothing would work with CUDA/HIP if `unsigned int` is not 4 bytes.

Reply:
The CUDA and HIP ecosystem require that `sizeof(int)==4`. Seeing it from the C++ standard view you are right but we do not guard it in many other places.